src/metabase/query_processor.clj

```diff
@@ -221,29 +233,21 @@
          query
          args))
 
-(def ^:private ^:dynamic *preprocessing-level* 1)
-
-(def ^:private ^:const max-preprocessing-level 20)
-
-(defn- preprocess-query [query context]
-  (binding [*preprocessing-level* (inc *preprocessing-level*)]
-    ;; record the number of recursive preprocesses taking place to prevent infinite preprocessing loops.
-    (log/tracef "*preprocessing-level*: %d" *preprocessing-level*)
-    (when (>= *preprocessing-level* max-preprocessing-level)
-      (throw (ex-info (str (tru "Infinite loop detected: recursively preprocessed query {0} times."
-                                max-preprocessing-level))
-               {:type error-type/qp})))
-    (process-query-sync query context)))
-
-(defn query->preprocessed
-  "Return the fully preprocessed form for `query`, the way it would look immediately before `mbql->native` is called.
-  Especially helpful for debugging or testing driver QP implementations."
+(defn preprocess
+  "Return the fully preprocessed form for `query`, the way it would look immediately
+  before [[mbql-to-native/mbql->native]] is called."
+  [query]
+  (let [qp (qp.reducible/combine-middleware
+            (conj (vec around-middleware)
+                  prevent-infinite-recursive-preprocesses/prevent-infinite-recursive-preprocesses)
+            (fn [query _rff _context]
+              (preprocess* query)))]
+    (qp query nil nil)))
```

Comment:
This feels weird to me. And also like the old world we got rid of. I'm surprised to see that we still need `_rff` and `_context`.

Reply:
I understand now. It's a shame we still need some of the `around-middleware` to perform this. It would be lovely if `preprocess` was just `((apply comp preprocess-middleware)) query)`